File: Tests/UnitTestImpromptuInterface/Generics.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Dynamic;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using ImpromptuInterface;

#if !SELFRUNNER
using NUnit.Framework;
#endif


#if SILVERLIGHT
namespace UnitTestImpromptuInterface.Silverlight
#else
namespace UnitTestImpromptuInterface
#endif
{
  [TestFixture]
    public class Generics : Helper
    {
        [Test]
        public void TestGenericMeth()
        {
       
            GenericMethHelper(3, "3");
            GenericMethHelper(4, "4");
            GenericMethHelper(true, "True");

            GenericMethHelper2(3);
            GenericMethHelper2(4);
            GenericMethHelper2(true);
            GenericMethHelper2("test'");
        }

        private void GenericMethHelper<T>(T param, string expected)
        {
            dynamic tNew = new ExpandoObject();
            tNew.Action = new Func<T, string>(it => it.ToString());
            IGenericMeth tActsLike = Impromptu.ActLike<IGenericMeth>(tNew);

            Assert.AreEqual(expected, tActsLike.Action(param));
        }

        private void GenericMethHelper2<T>(T param)
        {
            dynamic tNew = new ExpandoObject();
            tNew.Action2 = new Func<T, T>(it => it);
            IGenericMeth tActsLike = Impromptu.ActLike<IGenericMeth>(tNew);

            Assert.AreEqual(param, tActsLike.Action2(param));
        }


        [Test]
        public void TestGenericType()
        {

            GenericHelper(3, "3");
            GenericHelper(4, "4");
            GenericHelper(true, "True");
        }

        private void GenericHelper<T>(T param, string expected)
        {
            dynamic tNew = new ExpandoObject();
            tNew.Funct = new Func<T, string>(it => it.ToString());
            IGenericType<T> tActsLike = Impromptu.ActLike<IGenericType<T>>(tNew);

            Assert.AreEqual(expected, tActsLike.Funct(param));
        }

       [Test]
        public void TestGenericTypeConstraints()
        {

            var tObj = new Object();
            GenericHelperConstraints(tObj, tObj.ToString());
        }

        private void GenericHelperConstraints<T>(T param, string expected) where T : class
        {
            dynamic tNew = new ExpandoObject();
            tNew.Funct = new Func<T, string>(it => it.ToString());
            var tActsLike = Impromptu.ActLike<IGenericTypeConstraints<T>>(tNew);

            Assert.AreEqual(expected, tActsLike.Funct(param));
        }



        [Test]
        public void TestConstraintsMethGeneric()
        {
            var tObj = new Object();
            GenericMethConstraintsHelper(tObj, tObj.ToString());
            var tTest = Tuple.Create(new Uri(@"http://google.com"));
            GenericMethConstraintsHelper2(tTest, tTest.ToString());
        }

        private void GenericMethConstraintsHelper<T>(T param, string expected) where T : class
        {
            dynamic tNew = new ExpandoObject();
            tNew.Action = new Func<T, string>(it => it.ToString());
            var tActsLike = Impromptu.ActLike<IGenericMethWithConstraints>(tNew);

            Assert.AreEqual(expected, tActsLike.Action(param));
        }

        private void GenericMethConstraintsHelper2<T>(T param, string expected) where T : IComparable

        {
            dynamic tNew = new ExpandoObject();
            tNew.Action2 = new Func<T, string>(it => it.ToString());
            var tActsLike = Impromptu.ActLike<IGenericMethWithConstraints>(tNew);

            Assert.AreEqual(expected, tActsLike.Action2(param));
        }
    }
}

File: Tests/UnitTestImpromptuInterface.VBNet/Generics.vb
Option Strict Off

Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Dynamic
Imports System.Linq
Imports System.Runtime.Serialization
Imports System.Text
Imports NUnit.Framework
Imports ImpromptuInterface

Namespace VBNET


    <TestFixture()> _
    Public Class Generics
        Inherits Helper

        Delegate Function Test(Of T)() As String
        
        <Test()> _
        Public Sub TestFunctionWithGenericParameter()
            GenericMethHelper(3, "3")
            GenericMethHelper(4, "4")
            GenericMethHelper(True, "True")
        End Sub
        
        <Test()> _
        Public Sub TestFunctionWithGenericParameterAndReturnValue()
            GenericMethHelper2(3)
            GenericMethHelper2(4)
            GenericMethHelper2(True)
            GenericMethHelper2("test'")
        End Sub

        <Test()> _
        Public Sub TestFunctionWithGenericArgument()
            GenericMethHelper3(Of String)(GetType(String).Name)
        End Sub

        Private Sub GenericMethHelper(Of T)(param As T, expected As String)
            Dim tNew As Object = New ExpandoObject()
            tNew.Action = New Func(Of T, String)(Function(it) it.ToString())
            Dim tActsLike As IGenericMeth = Impromptu.ActLike(Of IGenericMeth)(tNew)

            Assert.AreEqual(expected, tActsLike.Action(param))
        End Sub

        Private Sub GenericMethHelper2(Of T)(param As T)
            Dim tNew As Object = New ExpandoObject()
            tNew.Action2 = New Func(Of T, T)(Function(it) it)
            Dim tActsLike As IGenericMeth = Impromptu.ActLike(Of IGenericMeth)(tNew)

            Assert.AreEqual(param, tActsLike.Action2(param))
        End Sub

        Private Sub GenericMethHelper3(Of T)(expected As String)
            Dim tNew As Object = New ExpandoObject()
            tNew.Action3 = CType(AddressOf Action3Stub(Of T), Test(Of T))
            Dim tActsLike As IGenericMeth = Impromptu.ActLike(Of IGenericMeth)(tNew)

            Assert.AreEqual(expected, tActsLike.Action3(Of T))
        End Sub

        Private Function Action3Stub(Of T)() As String
            Return GetType(T).Name
        End Function

        <Test()> _
        Public Sub TestGenericType()

            GenericHelper(3, "3")
            GenericHelper(4, "4")
            GenericHelper(True, "True")
        End Sub

        Private Sub GenericHelper(Of T)(param As T, expected As String)
            Dim tNew As Object = New ExpandoObject()
            tNew.Funct = New Func(Of T, String)(Function(it) it.ToString())
            Dim tActsLike As IGenericType(Of T) = Impromptu.ActLike(Of IGenericType(Of T))(tNew)

            Assert.AreEqual(expected, tActsLike.Funct(param))
        End Sub

        <Test()> _
        Public Sub TestGenericTypeConstraints()

            Dim tObj = New [Object]()
            GenericHelperConstraints(tObj, tObj.ToString())
        End Sub

        Private Sub GenericHelperConstraints(Of T As Class)(param As T, expected As String)
            Dim tNew As Object = New ExpandoObject()
            tNew.Funct = New Func(Of T, String)(Function(it) it.ToString())
            Dim tActsLike = Impromptu.ActLike(Of IGenericTypeConstraints(Of T))(tNew)

            Assert.AreEqual(expected, tActsLike.Funct(param))
        End Sub



        <Test()> _
        Public Sub TestConstraintsMethGeneric()
            Dim tObj = New [Object]()
            GenericMethConstraintsHelper(tObj, tObj.ToString())

        End Sub

        Private Sub GenericMethConstraintsHelper(Of T As Class)(param As T, expected As String)
            Dim tNew As Object = New ExpandoObject()
            tNew.Action = New Func(Of T, String)(Function(it) it.ToString())
            Dim tActsLike = Impromptu.ActLike(Of IGenericMethWithConstraints)(tNew)

            Assert.AreEqual(expected, tActsLike.Action(param))
        End Sub


    End Class
End Namespace